### PR TITLE
Hard negative mining for Retriever fine-tuning

### PR DESCRIPTION
**Pull request by vinay-raman**
## Description
Provides functionality to create training datasets for retriever customization

## Usage
1. Semantically cluster documents into partitions:
```python 
python3 repartition.py --input-dir=<input_directory> --hard-negative-mining-config=<your-config-file.yaml> --output-dir=<output-directory> --api-key=<your-api-key>
```
2. Mine hard negatives separately for each of the partitions
```python 
python3 mine_hard_negatives.py --input-dir=<output_directory_of step1>/clustered_dataset/ --hard-negative-mining-config=<your-config-file.yaml> --output-dir=<output-directory> --api-key=<your-api-key>
```
## Checklist

- [x] I am familiar with the [Contributing Guide](https://github.com/NVIDIA/NeMo-Curator/blob/main/CONTRIBUTING.md).
- [x] New or Existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: nemo_curator/modules/semantic_dedup/clusteringmodel.py [hunk 1]
```diff
@@ -124,7 +124,7 @@ def __call__(self, embeddings_dataset: DocumentDataset):
             )
 
         with performance_report_if_with_ts_suffix(self.profile_dir, "clustering-model"):
-            embeddings_df = embeddings_df[[self.id_col, self.embedding_col]]
+            # embeddings_df = embeddings_df[[self.id_col, self.embedding_col]]
```
Comment (ryantwolf) [inline nemo_curator/modules/semantic_dedup/clusteringmodel.py:127]: Why was this change made?

File: tutorials/nemo-retriever-synthetic-data-generation/config/config.yaml [hunk 2]
```diff
@@ -3,7 +3,7 @@ base_url: https://integrate.api.nvidia.com/v1
 api_key: "your api key here"
 
 # LLM Generator Module
-generator_model: mistralai/mixtral-8x22b-instruct-v0.1
+generator_model: nvdev/mistralai/mixtral-8x22b-instruct-v0.1
```
Comment (ryantwolf) [inline tutorials/nemo-retriever-synthetic-data-generation/config/config.yaml:6]: Revert

File: tutorials/nemo-retriever-synthetic-data-generation/config/config.yaml [hunk 3]
```diff
@@ -57,7 +57,7 @@ generator_user_prompt_template: |
 
 
 #Easiness filter (embedding model as judge)
-easiness_filter: nvidia/nv-embedqa-e5-v5
+easiness_filter: nvdev/nvidia/nv-embedqa-e5-v5
```
Comment (ryantwolf) [inline tutorials/nemo-retriever-synthetic-data-generation/config/config.yaml:60]: Revert

File: tutorials/nemo-retriever-synthetic-data-generation/config/hard-negative-mining-config.yaml [hunk 4]
```diff
@@ -0,0 +1,23 @@
+# HARD-NEGATIVE MINING parameters
+# base_url: "https://integrate.api.nvidia.com/v1"
+model_name: "sentence-transformers/all-MiniLM-L6-v2"
+# "intfloat/e5-large-unsupervised"
```
Comment (ryantwolf) [inline tutorials/nemo-retriever-synthetic-data-generation/config/hard-negative-mining-config.yaml:4]: Remove commented out config.

File: tutorials/nemo-retriever-synthetic-data-generation/config/hard-negative-mining-config.yaml [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+hard_negatives_to_mine: 4
+hard_neg_mining_algorithm: "topk_percpos"
+percpos: 0.95
+# max_hardness_threshold: 0.75
+# min_hardness_threshold: 0
+
+# SEMANTIC CLUSTERING parameters
+min_cluster_size: 50
+max_number_clusters: 200
+cluster_output_dir: "/home/viraman/qa_bug_fix/curator_outputs/clusters"
```
Comment (ryantwolf) [inline tutorials/nemo-retriever-synthetic-data-generation/config/hard-negative-mining-config.yaml:22]: Change these directories to the defaults in your config.py.

File: tutorials/nemo-retriever-synthetic-data-generation/retriever_hardnegative_miner.py [hunk 6]
```diff
@@ -0,0 +1,304 @@
...
+    ) -> DocumentDataset:
+        df = dataset.df
+        n_data = df.compute().shape[0]  # number of row items
+        n_clusters = int(np.floor(n_data / self.min_cluster_size) + 1)
+        n_clusters = min(n_clusters, self.max_number_clusters)
+        print("Number of clusters used = {}".format(n_clusters))
+        assert "doc_id" not in df.columns
+        df["embeddings"] = ""  # refers to document embeddings
+        df = df.map_partitions(self._get_doc_embeddings, meta=df)
+        # df = dd.from_pandas(pdf)
```
Comment (ryantwolf) [inline tutorials/nemo-retriever-synthetic-data-generation/retriever_hardnegative_miner.py:111]: Remove commented out code.

File: tutorials/nemo-retriever-synthetic-data-generation/retriever_hardnegative_miner.py [hunk 7]
```diff
@@ -0,0 +1,304 @@
...
+def create_nim_client(base_url, api_key):
+    openai_client = OpenAI(base_url=base_url, api_key=api_key)
+    return openai_client
+
+
+def create_hf_model(model_name_or_path):
+    return SentenceTransformer(model_name_or_path, trust_remote_code=True)
+
+
+class HardNegativeMiner:
```
Comment (ryantwolf) [inline tutorials/nemo-retriever-synthetic-data-generation/retriever_hardnegative_miner.py:46]: Add docstring.

File: tutorials/nemo-retriever-synthetic-data-generation/retriever_hardnegative_miner.py [hunk 8]
```diff
@@ -0,0 +1,304 @@
...
+        df["query_embed"] = ""
+        df["min_pos_score"] = ""
+
+        with ProgressBar(dt=1):
+            df = df.map_partitions(self._process_partition, meta=df).compute()
+
+        df = df.rename(columns={"documents": "pos_doc"})
+        df = df[["question", "pos_doc", "neg_doc"]]
+
+        return DocumentDataset(dd.from_pandas(df))
```
Comment (ryantwolf) [inline tutorials/nemo-retriever-synthetic-data-generation/retriever_hardnegative_miner.py:180]: Why are you doing `dd.from_pandas` here? I don't see where `df` becomes a pandas dataframe. `df = df.to_backend("pandas")` makes it such that `df` is still a Dask dataframe with a pandas backend, but it's still a distributed dataframe.

File: tutorials/nemo-retriever-synthetic-data-generation/retriever_hardnegative_miner.py [hunk 9]
```diff
@@ -0,0 +1,304 @@
...
+        print("Number partitions in dataset = {}".format(df.npartitions))
+
+        df["neg_doc_scores"] = ""
+        df["neg_doc"] = ""
+        df["doc_embed"] = ""
+        df["query_embed"] = ""
+        df["min_pos_score"] = ""
+
+        with ProgressBar(dt=1):
+            df = df.map_partitions(self._process_partition, meta=df).compute()
```
Comment (ryantwolf) [inline tutorials/nemo-retriever-synthetic-data-generation/retriever_hardnegative_miner.py:175]: Do not call compute. It's up to the user when to call compute outside of the modules.

File: tutorials/nemo-retriever-synthetic-data-generation/config/hard-negative-mining-config.yaml [hunk 10]
```diff
@@ -0,0 +1,23 @@
+# HARD-NEGATIVE MINING parameters
+# base_url: "https://integrate.api.nvidia.com/v1"
+model_name: "sentence-transformers/all-MiniLM-L6-v2"
+# "intfloat/e5-large-unsupervised"
+# "nvdev/nvidia/llama-3.2-nv-embedqa-1b-v1"
+# nvdev/nvidia/nv-embedqa-e5-v5"
+# "nvdev/nvidia/llama-3.2-nv-embedqa-1b-v1"
+model_type: "hf"
+query_prefix: "query:"
```
Comment (ryantwolf) [inline tutorials/nemo-retriever-synthetic-data-generation/config/hard-negative-mining-config.yaml:9]: Is there a reason why the defaults in the dataclass don't match the defaults in the yaml?

File: tutorials/nemo-retriever-synthetic-data-generation/retriever_hardnegative_miner.py (hunk 7 above)
Comment (ryantwolf) [inline tutorials/nemo-retriever-synthetic-data-generation/retriever_hardnegative_miner.py:46]: This seems like a general enough module that we could move it into `nemo_curator/` and have this tutorial just show how to use it, what do you think? If we did, you'd also need to inherit from `nemo_curator.modules.base.BaseModule`.

File: tutorials/nemo-retriever-synthetic-data-generation/retriever_hardnegative_miner.py [hunk 11]
```diff
@@ -0,0 +1,304 @@
...
+                load_object_kwargs={"base_url": self.base_url, "api_key": self.api_key},
+            )
+            df_p["doc_embed"] = df_p["documents"].map(
+                lambda pgs: [self._get_nim_embedding(t, "passage") for t in pgs]
+            )
+            df_p["query_embed"] = df_p["question"].map(
+                lambda x: self._get_nim_embedding(x, "query")
+            )
+        elif self.model_type == "hf":
+            self.hf_model = load_object_on_worker(
```
Comment (ryantwolf) [inline tutorials/nemo-retriever-synthetic-data-generation/retriever_hardnegative_miner.py:197]: @VibhuJawa should this be using crossfit?

**Review by ryantwolf — CHANGES_REQUESTED** [review]
You should update the README for this too. It should answer the following questions:
1. What is hard negative mining?
2. When should I use hard negative mining?
3. How does this version of hard negative mining work?
4. How can I use these scripts?

File: tutorials/nemo-retriever-synthetic-data-generation/retriever_hardnegative_miner.py [hunk 12]
```diff
@@ -0,0 +1,304 @@
...
+                lambda pgs: [
+                    self._get_hf_embedding(t, self.passage_prefix) for t in pgs
+                ]
+            )
+        return p_df
+
+    def _groupby_question(self, pdf):
+        pdf2 = pdf.groupby("question").agg({"documents": set})
+        pdf2["documents"] = pdf2["documents"].map(lambda x: list(x))
+        del pdf
```
Comment (ryantwolf) [inline tutorials/nemo-retriever-synthetic-data-generation/retriever_hardnegative_miner.py:157]: Why call this `del`?

File: tutorials/nemo-retriever-synthetic-data-generation/retriever_hardnegative_miner.py [hunk 13]
```diff
@@ -0,0 +1,304 @@
...
+        return pdf2
+
+    def __call__(self, dataset: DocumentDataset) -> DocumentDataset:
+
+        df = dataset.df
+        df = df.to_backend("pandas")
+        df = df[["question", "documents"]]
+        df = df.map_partitions(self._groupby_question).reset_index()
```
Comment (ryantwolf) [inline tutorials/nemo-retriever-synthetic-data-generation/retriever_hardnegative_miner.py:165]: Are you confident that all the questions will be in the same partition, or should you do a shuffle first?

File: tutorials/nemo-retriever-synthetic-data-generation/retriever_hardnegative_miner.py [hunk 14]
```diff
@@ -0,0 +1,304 @@
...
+        if cfg.query_prefix:
+            self.query_prefix = cfg.query_prefix
+        if cfg.hard_neg_mining_algorithm:
+            self.hard_neg_mining_algorithm = cfg.hard_neg_mining_algorithm
+        else:
+            print(
+                "hard negative mining algorithm not mentioned in config, using default"
+            )
+            self.hard_neg_mining_algorithm = "topk_percpos"
+        if self.hard_neg_mining_algorithm == "topk_percpos":
```
Comment (ryantwolf) [inline tutorials/nemo-retriever-synthetic-data-generation/retriever_hardnegative_miner.py:71]: Can we abstract the algorithms and their arguments into a separate class? Follow a similar pattern to the [common crawl extraction algorithms](https://github.com/NVIDIA/NeMo-Curator/blob/main/nemo_curator/download/commoncrawl.py#L74). We should have one base `HardNegativeMiningAlgorithm` class that defines a single method with a common signature across all subclasses. Instances of those subclasses can be passed in as arguments to this class.

If you feel like too much of the state of the `HardNegativeMiner` class would need to be passed to this algorithm, then we could instead consider making `HardNegativeMiner` an ABC that these specific methods inherit from. Let me know what you think.

File: tutorials/nemo-retriever-synthetic-data-generation/retriever_hardnegative_miner.py [hunk 15]
```diff
@@ -0,0 +1,304 @@
...
+        return normalize_token(HardNegativeMiner)
+
+    def assign_ids(self, partition):
+        return partition.assign(doc_id=np.arange(len(partition)) + partition.index[0])
+
+    def repartition_semantic_similarity(
+        self, dataset: DocumentDataset
+    ) -> DocumentDataset:
+        df = dataset.df
+        n_data = df.compute().shape[0]  # number of row items
```
Comment (VibhuJawa) [inline tutorials/nemo-retriever-synthetic-data-generation/retriever_hardnegative_miner.py:104]: This will make sure we dont do double computation here and only pull in shape 

```suggestion
        df = df.persist()
        n_data = df.shape[0].compute()
```

File: tutorials/nemo-retriever-synthetic-data-generation/retriever_hardnegative_miner.py [hunk 16]
```diff
@@ -0,0 +1,304 @@
...
+            df = df.map_partitions(self._process_partition, meta=df).compute()
+
+        df = df.rename(columns={"documents": "pos_doc"})
+        df = df[["question", "pos_doc", "neg_doc"]]
+
+        return DocumentDataset(dd.from_pandas(df))
+
+    def _process_partition(self, df_p: pd.DataFrame):
+
+        if self.model_type == "nvidia":
```
Comment (VibhuJawa) [inline tutorials/nemo-retriever-synthetic-data-generation/retriever_hardnegative_miner.py:184]: Can we change this model_type name to `nim` or `open_ai client` , model_type="nvidia" is confusing

```suggestion
        if self.model_type == "nim":
```

File: tutorials/nemo-retriever-synthetic-data-generation/retriever_hardnegative_miner.py (hunk 11 above)
Comment (VibhuJawa) [reply tutorials/nemo-retriever-synthetic-data-generation/retriever_hardnegative_miner.py:197]: Yes, it can be, crossfit has support for these kind of embeddings . This will need some re-work to make that happen. I am not sure about time constraints at play here

**Review by VibhuJawa — CHANGES_REQUESTED** [review]
Took an intial look, will take a deeper look in a bit more time.
